Add Support for Editing Unique Names

GCRA101 opened the pull request:
### Issues addressed by this PR

 Fixes #514


<img width="1660" height="796" alt="Grasshopper Script View" src="https://github.com/user-attachments/assets/849b04be-e13f-45e9-b905-ae020aca8ac1" />

<img width="1510" height="926" alt="ETABS Model View" src="https://github.com/user-attachments/assets/02b8fea4-bb17-474d-a336-4e2331e536f2" />


ETABS Toolkit now allows to push/pull and update objects (i.e. bars, nodes, panels, rigid links and openings) handling corresponding assignments to ETABS Unique Name via BHoM Name.


 ### Test files
Grasshopper Files
- UniqueNames Test Script: 
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23518-AddSupportForEditingUniqueNames?csf=1&web=1&e=DFqPCh
- Groups Test Script:
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23518-AddSupportForEditingUniqueNames?csf=1&web=1&e=DFqPCh

 ### Changelog
- Add Support for UniqueName Assignment in Push/Pull/Update of Bars
- Add Support for UniqueName Assignment in Push/Pull/Update of Nodes
- Add Support for UniqueName Assignment in Push/Pull/Update of Panels
- Add Support for UniqueName Assignment in Push/Pull/Update of Links
- Add Support for UniqueName Assignment in Push/Pull/Update of Openings

GCRA101 committed:
Add Unique Name Assignment in Push/Pull of Bars

GCRA101 committed:
Add Unique Name Assignment in Push/Pull of Nodes

GCRA101 committed:
Add Unique Name Assignment in Push/Pull of Panels

GCRA101 committed:
Add Unique Name Assignment in Push/Pull of Openings

GCRA101 committed:
Fix bug in unique names assignment for bars, panels, openings and nodes

GCRA101 committed:
Add Unique Name Assignment in Push/Pull of Rigid Links

GCRA101 committed:
Add utility method GetBhomNameFromEtabsId() in Pull Methods for Bars, Links, Nodes, Panels and Openings

GCRA101 committed:
Add Unique Name Assignment in Update of Bars

GCRA101 committed:
Add Assignment of UniqueName in Update of Nodes

GCRA101 committed:
Add Assignment of UniqueName in Update of Panels

GCRA101 commented:
@BHoMBot check installer

Pritesh2024 commented:
Script tested on a real project.

albertbg028 reviewed (dismissed):
New feature works as intended.

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@peterjamesnugent, @Chrisshort92,
fyi, I think this one is ready for merge. Functionality tested by both Pritish and Luis.
All BHoMBot checks have passed.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Bar.cs:
```diff
@@ -95,15 +95,20 @@ private bool CreateObject(Bar bhBar)
 
-            ETABSId etabsIdFragment = new ETABSId { Id = name };
+            // Assign the Unique Name to the ETABS Element
+            string newName = SetUniqueName(bhBar, name);
```
Does this not mean every single `Bar` element pushed to ETABS will have a unique name? Is this the aim of the issue? I thought it was to add support for names to ETABS in which case elements can have the same names?

GCRA101 replied:
@peterjamesnugent, thanks for the comment.
ETABS doesn't allow to have more than one element with the same Unique Name, but Unique Name is the only feature that allows to provide names for filtering/grouping/classification of ETABS elements.
With this new feature the user is allowed to assign whatever name they need the ETABS elements without the risk of crashing the model since the given name gets combined with a portion of the GUID of the element (this makes essentially impossible ending up with any duplicate unique name).
In a nutshell, this allows to assign the same name to multiple objects while abiding by the naming restrictions of ETABS.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Link.cs:
```diff
@@ -106,6 +114,42 @@ private bool CreateObject(LinkConstraint bhLinkConstraint)
         }
 
         /***************************************************/
+
+        [Description("Concatenates the last 7 characters of the ETABS Element GUID and the Link Name to get the Unique Name to assign to the ETABS Element.")]
+        private List<string> SetUniqueName(RigidLink bhLink, List<string> names)
+        {
+
+            int ret01, ret02;
+            string guid = null;
+            string tempLinkName = "";
+            List<string> newLinkNames = new List<string>();
+
+            foreach (string name in names) {
+
+                tempLinkName = "";
+                ret01 = m_model.LinkObj.GetGUID(name, ref guid);
+
+                if (bhLink.Name == "")
+                {
+                    tempLinkName = guid.Substring(guid.Length - 7);
+                }
+                else
+                {
+                    tempLinkName = guid.Substring(guid.Length - 7) + "::" + bhLink.Name;
+                }
+
+                ret02 = m_model.LinkObj.ChangeName(name, tempLinkName);
+
+                newLinkNames.Add(tempLinkName);
+
+                if (!(ret01 == 0 && ret02 == 0)) return null;
+
+            }
+
+            return newLinkNames;
+        }
```
If you are intent on giving every element a unique name (see comment above) - can you not just make a single method with a `BHoMObject` input given it will have the properties of `Name` and `BHoMGuid`?

GCRA101 replied:
@peterjamesnugent,
thanks for the comment. Fair enough. I'll replace it with a generic method working with BHoMObject.

GCRA101 replied:
@peterjamesnugent, the only type of object that needs to be treated separately is the RigidLink object type. 

I've created so a generic method dealing with BHoMObject and that applies for Bars, Nodes, Panels and Openings, while the RigidLinks are treated with their own private method using its own logic.

I'll push the updated commit shortly.

<img width="1711" height="652" alt="image" src="https://github.com/user-attachments/assets/048934a0-8c92-46c3-8270-7d4e48106b7b" />

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Node.cs:
```diff
@@ -106,6 +115,34 @@ private bool SetObject(Node bhNode, string name)
         /***************************************************/
+
+        [Description("Concatenates the last 7 characters of the ETABS Element GUID and the Node Name to get the Unique Name to assign to the ETABS Element.")]
+        private string SetUniqueName(Node bhNode, string name)
```
As above.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Opening.cs:
```diff
@@ -84,31 +87,64 @@ private bool CreateObject(Opening bhOpening)
         /***************************************************/
 
+        [Description("Concatenates the last 7 characters of the ETABS Element GUID and the Opening Name to get the Unique Name to assign to the ETABS Element.")]
+        private string SetUniqueName(Opening bhOpening, string name)
```
As above.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Panel.cs:
```diff
@@ -198,6 +207,31 @@ private static void NonLinearEdgesCheck(List<Edge> edges)
+        /***************************************************/
+
+        [Description("Concatenates the last 7 characters of the ETABS Element GUID and the Panel Name to get the Unique Name to assign to the ETABS Element.")]
+        private string SetUniqueName(Panel bhPanel, string name)
```
As above.

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/_Read.cs:
```diff
@@ -186,6 +186,20 @@ private static List<string> FilterIds(IEnumerable<string> ids, IEnumerable<strin
+        [Description("Extracts the String Name of the BHoM object from the corresponding ETABS Unique Name.")]
+        public string GetBhomNameFromEtabsId(string id)
+        {
+            if (id.Contains("::"))
```
This is all dependent on the unique name function above. Not sure why we need to have all names of each element unique.

GCRA101 replied:
See comment above https://github.com/BHoM/ETABS_Toolkit/pull/518#discussion_r2602061477

peterjamesnugent commented on Etabs_Adapter/CRUD/Update/Node.cs:
```diff
@@ -53,85 +58,124 @@ private bool UpdateObjects(IEnumerable<Node> nodes)
+        /***************************************************/
+
+        [Description("Concatenates the last 7 characters of the ETABS Element GUID and the Node Name to get the Unique Name to assign to the ETABS Element.")]
+        private bool UpdateUniqueName(Node bhNode)
```
Does this method need to be unique to a `Node`? Can it not be a generic `BHoMObject` method?

peterjamesnugent reviewed (changes requested):
Not sure I agree with the need for unique names for each element (unless I have misunderstood?). Also, a few of the methods could be consolidated given they just use `Name` and `Guid` properties.

samuelbasimalla commented:
> Not sure I agree with the need for unique names for each element (unless I have misunderstood?). Also, a few of the methods could be consolidated given they just use `Name` and `Guid` properties.

@peterjamesnugent I discussed the use case for editing unique names with @GCRA101 earlier. 
The ability to "assign groups" operate at higher level of hierarchy. "Editing Unique names" operate at a level below.

Example 1: All the similar Pad foundations with similar demands can be assigned a group. However, each pad foundation, which is usually modelled as panels is further subdivided into smaller panels (meshing requirements) which currently have randomly generated unique names. If there is a feature to assign unique names, we can do so in a way that each of the sub-panels identify with the a Pad. This basically helps in post-processing the results for element design.

Example 2: Same as above for bar elements which are sub-divided into smaller bar elements for meshing, but for design where we would need full buckling lengths, which is where unique names can help. 

Hope I am clear.

GCRA101 committed:
Refactor the SetUniqueName() Method in the Create/ folder.

Replace all the SetUniqueName private methods defined for bars, nodes, panels and openings with a single method taking as input a BHoMObject.
Taking advantage of polymorphism, we can minimize the repetition of code within the Toolkit.

GCRA101 committed:
Refactor the UpdateUniqueName() Method in the Update/ folder.

Replace all the UpdateUniqueName private methods defined for bars, nodes, panels and openings with a single method taking as input a BHoMObject.
Taking advantage of polymorphism, we can minimize the repetition of code within the Toolkit.

GCRA101 commented:
@peterjamesnugent,

thanks again for your comment.
You can find the updated code incorporating your comments in the latest commit: https://github.com/BHoM/ETABS_Toolkit/pull/518/commits/2529a92a125d29a3ef43d564d3c2e39afd98be23

peterjamesnugent commented:
As discussed with @GCRA101 and @samuelbasimalla, this will be closed to allow https://github.com/BHoM/ETABS_Toolkit/pull/516 to take priority. 

The unique name attribute using part of the id doesn't help the user when they could just use `Tags` to filter their objects whether applying loads, or pulling results. The way I see the `Pull` working is:
1. User pulls all objects e.g. `Bar`, `Panel` etc.
2. User uses the `FilterRequest in https://github.com/BHoM/BHoM/blob/c7a25f7799782f9025ead7f532b04a6cd33c7ff0/Data_oM/Requests/FilterRequest.cs#L28 to specify the `Tags`
3. User makes user of https://github.com/BHoM/BHoM_Engine/blob/ee49f81ea511d7ee0c66a3c13d0d73f1a3677bd8/Data_Engine/Compute/FilterData.cs#L37 to filter the data as needed.

This way, the user only needs to know the tags assigned, and doesn't need to do extra work to determine the unique name assigned by the guid. For example, if I wanted to pull 100 `PadFoundation` to design, I would either need to retrieve that from the unique names in ETABS, or break down a string using the `Name` and `BHoM_Guid` attribute whilst having knowledge of how many letters that string uses which the average user won't be aware of.

Whereas, using the `Tags`, the user can just use the in built `FilterRequest` and `FilterData` objects and methods to breakdown their data as they see fit. The tags are known, and importantly `Tags` are many-to-one whereas these unique names are one-to-one.

GCRA101 committed:
Merge branch 'develop' into ETABS_Toolkit-#514-AddSupportForEditingUniqueNames

GCRA101 committed:
Fix syntax error in _Update.cs

albertbg028 reviewed (approved):
Tested bars, panels, and nodes elements using the provided script. The script performed as expected, appending the custom naming to the existing unique names for each element.
